Домашнее задание №17. Кручинин С.В.
Задача. Оценка книги. Масштаб, Lasso и Ridge, в том числе квадрат возраста.
Запускать из папки дз17.


In [1]:
import pandas as pd
from sklearn.linear_model import Lasso, Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler

df = pd.read_csv('priznaki.csv')
df = df[df['otsenka'].notna()].copy()
df['vozrast2'] = df['vozrast'] ** 2
print('строк', len(df))

def check(name, model, cols):
    X = df[cols]
    y = df['otsenka']
    Xtr, Xte, ytr, yte = train_test_split(X, y, test_size=0.33, random_state=42)
    scaler = StandardScaler()
    Xtr = scaler.fit_transform(Xtr)
    Xte = scaler.transform(Xte)
    model.fit(Xtr, ytr)
    pred = model.predict(Xte)
    print(name, cols)
    print('w0', round(float(model.intercept_), 3), 'w', [round(float(v), 3) for v in model.coef_])
    print('MSE', round(mean_squared_error(yte, pred), 3),
          'MAE', round(mean_absolute_error(yte, pred), 3),
          'R2', round(r2_score(yte, pred), 3))

for kind, cls in (('Lasso', Lasso), ('Ridge', Ridge)):
    check(kind, cls(alpha=0.1), ['vozrast'])
    check(kind, cls(alpha=0.1), ['tema_sovpala'])
    check(kind, cls(alpha=0.1), ['vozrast', 'tema_sovpala'])
    check(kind, cls(alpha=0.1), ['vozrast', 'vozrast2', 'tema_sovpala'])
    check(kind, cls(alpha=0.1), ['vozrast', 'tema_sovpala', 'zhanr_sovpal', 'skok_brali'])


строк 81
Lasso ['vozrast']
w0 3.796 w [0.248]
MSE 1.164 MAE 0.931 R2 0.007
Lasso ['tema_sovpala']
w0 3.796 w [0.359]
MSE 1.001 MAE 0.843 R2 0.145
Lasso ['vozrast', 'tema_sovpala']
w0 3.796 w [0.209, 0.334]
MSE 0.942 MAE 0.832 R2 0.196
Lasso ['vozrast', 'vozrast2', 'tema_sovpala']
w0 3.796 w [0.209, 0.0, 0.334]
MSE 0.942 MAE 0.832 R2 0.196
Lasso ['vozrast', 'tema_sovpala', 'zhanr_sovpal', 'skok_brali']
w0 3.796 w [0.237, 0.391, -0.126, -0.0]
MSE 0.981 MAE 0.848 R2 0.163
Ridge ['vozrast']
w0 3.796 w [0.348]
MSE 1.199 MAE 0.945 R2 -0.024
Ridge ['tema_sovpala']
w0 3.796 w [0.458]
MSE 0.993 MAE 0.836 R2 0.152
Ridge ['vozrast', 'tema_sovpala']
w0 3.796 w [0.298, 0.423]
MSE 0.92 MAE 0.821 R2 0.215
Ridge ['vozrast', 'vozrast2', 'tema_sovpala']
w0 3.796 w [1.57, -1.285, 0.414]
MSE 0.993 MAE 0.85 R2 0.152
Ridge ['vozrast', 'tema_sovpala', 'zhanr_sovpal', 'skok_brali']
w0 3.796 w [0.376, 0.585, -0.356, 0.009]
MSE 1.069 MAE 0.895 R2 0.088


Ответ. Квадрат возраста Lasso обнулил. Лучше Ridge на возрасте и теме: MSE 0.92, R2 0.22.
